File: leetcode2/1easy/최은서/Q2843.js
/**
 * @param {number} low
 * @param {number} high
 * @return {number}
 */
var countSymmetricIntegers = function(low, high) {
    let cnt = 0;
    for (let i = low; i <= high; i++) { //O(n)
        if (i.toString().length % 2 !== 0) continue; //홀수 길이의 숫자는 대칭 정수가 아님
        const mid = i.toString().length / 2 //문자열의 길이를 2로 나눈 몫 (중간 인덱스)
        const left = i.toString().substring(0, mid) //substring(start, end) : start부터 end까지의 문자열을 반환
        const right = i.toString().substring(mid, i.length) //substring(start, end) : start부터 end까지의 문자열을 반환
        const sum = str => [...str].reduce((acc, ch) => acc + Number(ch), 0); //배열의 합을 구하는 sum 함수
        //string 을 배열로 펼치고, 각 요소를 숫자로 변환해서 (Number(ch)) 누적 합을 구함 (시작 : 0)
        if (sum(left) === sum(right)) cnt += 1
    }
    return cnt
    
};

File: leetcode2/1easy/최원준/Q2843.java
package Leetcode.최원준;

/*
1. 아이디어 :
- 대칭인지 확인하는 메서드 isSymmetrical을 구현한다.
    - 숫자 길이의 반 만큼 오른쪽과 왼쪽의 합을 구한다.
- low부터 high까지 반복하면서, 숫자의 길이가 짝수인 경우에만 isSymmetrical을 호출한다.

2. 시간복잡도 :
O( n * m )

3. 자료구조/알고리즘 :
- / -
 */

public class Q2843 {
    class Solution {
        public int isSymmetrical(int num) {
            int length = String.valueOf(num).length();
            int right = 0;
            for (int i=0; i<length/2; i++) {
                right += num % 10;
                num /= 10;
            }

            int left = 0;
            for (int i=0; i<length/2; i++) {
                left += num % 10;
                num /= 10;
            }
            return (left == right)? 1:0;
        }
        public int countSymmetricIntegers(int low, int high) {
            int ans = 0;
            for (int num = low; num <= high; num++) {
                if (String.valueOf(num).length() % 2 != 0) continue;
                ans += isSymmetrical(num);
            }
            return ans;
        }
    }
}
